File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/vecarith/measuresystem.py
import vecop

def basefactors(stepfactors):
    result = []
    for (i, stepfactor) in enumerate(stepfactors):
        result.append(reduce(lambda x, y: x * y, stepfactors[i + 1:], 1))
    return result

def basefactors2(stepfactors):
    def acc_basefactors2(acc, stepfactor):
        acc.append(acc[-1] * stepfactor)
        return acc

    return reduce(acc_basefactors2, stepfactors[::-1][:-1], [1])[::-1]

def base(quantityvec, basefactors):
    return vecop.dotproduct(quantityvec, basefactors)

def equal(quantityvec1, quantityvec2, basefactors):
    return base(quantityvec1, basefactors) == base(quantityvec2, basefactors)

def normalize(quantityvalue, basefactors):
    return vecop.dotdiv(quantityvalue, basefactors)

def add(quantityvec1, quantityvec2, basefactors):
    return normalize(base(quantityvec1, basefactors) + base(quantityvec2, basefactors), basefactors)

def scale(c, quantityvec, basefactors):
    return normalize(c * base(quantityvec, basefactors), basefactors)

File: DDD DSL/DDD DSL Design War Measurement System/Code/Java/src/fayelab/ddd/measure/vecarith/MeasureSystem.java
package fayelab.ddd.measure.vecarith;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class MeasureSystem
{
    private List<Integer> baseFactors;

    public MeasureSystem(List<Integer> stepFactors)
    {
        this.baseFactors = toBaseFactors(stepFactors);
    }

    static List<Integer> toBaseFactors(List<Integer> stepFactors)
    {
        int size = stepFactors.size();
        return IntStream.range(0, size)
                        .mapToObj(i -> stepFactors.subList(i + 1, size)
                                                  .stream()
                                                  .reduce(1, (acc, factor) -> acc * factor))
                        .collect(Collectors.toList());
    }

    public int base(List<Integer> quantityVec)
    {
        return VecOp.dotProduct(quantityVec, baseFactors);
    }

    public boolean equal(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return base(quantityVec1) == base(quantityVec2);
    }

    public List<Integer> normalize(int quantityValue)
    {
        return VecOp.dotDiv(quantityValue, baseFactors);
    }

    public List<Integer> add(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return normalize(base(quantityVec1) + base(quantityVec2));
    }

    public List<Integer> scale(int c, List<Integer> quantityVec)
    {
        return normalize(c * base(quantityVec));
    }
}
